### PR TITLE
Introducing Product type

### PR DESCRIPTION
**Pull request by alari**

### CODE REVIEW

File: types/src/main/scala/aqua/types/Type.scala [hunk 1]
```diff
@@ -12,12 +12,96 @@ sealed trait Type {
...
+    case ConsType(t, pt) => Some(t -> pt)
+    case _ => None
+  }
+
+  def toList: List[Type] = this match {
+    case ConsType(t, pt) => t :: pt.toList
+    case _ => Nil
+  }
+
+  def toLabelledList(prefix: String = "arg", index: Int = 0): List[(String, Type)] = this match {
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:35]: docs

File: types/src/main/scala/aqua/types/Type.scala [hunk 2]
```diff
@@ -12,12 +12,96 @@ sealed trait Type {
     import cats.syntax.partialOrder._
     this >= incoming
   }
+
+  def isInhabited: Boolean = true
+}
+
+// Product is an ordered list of types, optionally labelled
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:19]: what does it mean `ordered` and `labelled`? more description in docs to make it clear what it is

File: types/src/main/scala/aqua/types/Type.scala [hunk 3]
```diff
@@ -12,12 +12,96 @@ sealed trait Type {
...
+  }
+
+  def labelled(types: List[(String, Type)]): ProductType = types match {
+    case (l, h) :: t =>
+      ConsType.cons(l, h, ProductType.labelled(t))
+    case _ => NilType
+  }
+}
+
+sealed trait ConsType extends ProductType {
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:64]: docs

File: types/src/main/scala/aqua/types/Type.scala [hunk 4]
```diff
@@ -151,37 +237,61 @@ object Type {
     ) -1.0
     else NaN
 
+  private def cmpProd(l: ProductType, r: ProductType, contra: Boolean): Double = (l, r) match {
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:240]: docs. Maybe rename? I don't understand what does it mean `cmpProd`

File: types/src/main/scala/aqua/types/Type.scala (hunk 2 above)
Comment (DieMyst) [reply types/src/main/scala/aqua/types/Type.scala:19]: add that `ordered` is for equality between types

File: types/src/main/scala/aqua/types/Type.scala (hunk 4 above)
Comment (DieMyst) [reply types/src/main/scala/aqua/types/Type.scala:240]: maybe `compare*`?

File: types/src/main/scala/aqua/types/Type.scala [hunk 5]
```diff
@@ -135,7 +221,7 @@ object Type {
         case _ => NaN
       }
 
-  private def cmpProd(lf: NonEmptyMap[String, Type], rf: NonEmptyMap[String, Type]): Double =
+  private def cmpStruct(lf: NonEmptyMap[String, Type], rf: NonEmptyMap[String, Type]): Double =
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:224]: docs and rename

File: types/src/main/scala/aqua/types/Type.scala [hunk 6]
```diff
@@ -151,37 +237,61 @@ object Type {
...
+      else {
+        val tailCmp = cmpProd(ltail, rtail, contra)
+        // If one is >, and another eq, it's >, and vice versa
+        if (headCmp >= 0 && tailCmp >= 0) 1.0
+        else if (headCmp <= 0 && tailCmp <= 0) -1.0
+        else NaN
+      }
+  }
+
   private def cmp(l: Type, r: Type): Double =
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:261]: rename to `compare`

File: types/src/main/scala/aqua/types/Type.scala [hunk 7]
```diff
@@ -100,21 +184,23 @@ case class OptionType(element: Type) extends BoxType {
...
-case class ProductType(name: String, fields: NonEmptyMap[String, Type]) extends DataType {
+// Struct is an unordered collection of labelled types
+case class StructType(name: String, fields: NonEmptyMap[String, Type]) extends DataType {
 
   override def toString: String =
     s"$name{${fields.map(_.toString).toNel.toList.map(kv => kv._1 + ": " + kv._2).mkString(", ")}}"
 }
 
-case class ArrowType(args: List[Type], res: Option[Type]) extends Type {
+case class ArrowType(domain: ProductType, codomain: ProductType) extends Type {
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:194]: docs
and i don't think that it is a good idea to rename arguments to `domain` and `codomain`

File: backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala [hunk 8]
```diff
@@ -113,12 +116,9 @@ case class JavaScriptFunc(func: FuncCallable) {
...
   def argsToTs(at: ArrowType): String =
-    at.args.zipWithIndex
-      .map(_.swap)
-      .map(kv => "arg" + kv._1)
-      .mkString(", ")
+    at.domain.toLabelledList().map(_._1).mkString(", ")
 
   def argsCallToJs(at: ArrowType): String =
-    at.args.zipWithIndex.map(_._2).map(idx => s"args[$idx]").mkString(", ")
+    at.domain.toList.zipWithIndex.map(_._2).map(idx => s"args[$idx]").mkString(", ")
```
Comment (DieMyst) [inline backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala:122]: is it `toLabelledList`?

File: types/src/main/scala/aqua/types/Type.scala [hunk 9]
```diff
@@ -100,21 +184,23 @@ case class OptionType(element: Type) extends BoxType {
...
+
+  def args: List[Type] = domain.toList
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:196]: lazy val

File: types/src/main/scala/aqua/types/Type.scala [hunk 10]
```diff
@@ -100,21 +184,23 @@ case class OptionType(element: Type) extends BoxType {
...
+  def res: Option[Type] = codomain.uncons.map(_._1)
```
Comment (DieMyst) [inline types/src/main/scala/aqua/types/Type.scala:197]: lazy val

File: backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala [hunk 11]
```diff
@@ -113,12 +116,9 @@ case class JavaScriptFunc(func: FuncCallable) {
 object JavaScriptFunc {
 
   def argsToTs(at: ArrowType): String =
-    at.args.zipWithIndex
-      .map(_.swap)
-      .map(kv => "arg" + kv._1)
-      .mkString(", ")
+    at.domain.toLabelledList().map(_._1).mkString(", ")
```
Comment (DieMyst) [inline backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala:119]: maybe use some helpers here to hide `domain` stuff. We already have `args` and `res` helper functions

File: backend/ts/src/main/scala/aqua/backend/ts/TypeScriptFunc.scala [hunk 12]
```diff
@@ -11,7 +11,7 @@ case class TypeScriptFunc(func: FuncCallable) {
   import TypeScriptFunc._
 
   def argsTypescript: String =
-    func.args.args.map(ad => s"${ad.name}: " + typeToTs(ad.`type`)).mkString(", ")
+    func.arrowType.domain.toLabelledList().map(ad => s"${ad._1}: " + typeToTs(ad._2)).mkString(", ")
```
Comment (DieMyst) [inline backend/ts/src/main/scala/aqua/backend/ts/TypeScriptFunc.scala:14]: avoid to use `domain` stuff in `backend` package

File: backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala (hunk 8 above)
Comment (alari) [reply backend/js/src/main/scala/aqua/backend/js/JavaScriptFunc.scala:122]: As you may see, no, it's just `toList`. Labels are skipped.

**Review by DieMyst — APPROVED** [review]

**Pull request merged**